### ISSUE TITLE
RcParams is fundamentally broken

### ISSUE THREAD

**Issue by timhoffm** [matplotlib/matplotlib#12576]
### Bug report

Anyone tried:

~~~
>>> import matplotlib
>>> len(matplotlib.rcParams)
0
~~~

A great example of shooting yourself in the foot with multiple inheritance. Essentially, we do

~~~
class RcParams(MutableMapping, dict):
    def __getitem__(self, key):
        return dict.__getitem__(self, key)

    def __setitem__(self, key, val):
        return dict.__setitem__(self, key, val)
~~~

but do not reimplement `__len__`. You can work out the consequences yourself.

### Proposed solution

Do not inherit `RcParams` from dict, and use a private data dict instead.

**Comment by timhoffm** [comment]
Fun fact: This results in `rcParams.clear()` to not have any effect.

**Comment by tacaswell** [comment]
Could we just implement `__len__` as a quicker fix?  I am curious of this ever worked....

**Comment by timhoffm** [comment]
Yes, implementing `__len__` is probably a workaround, and I'm fine with that. Don't have time to go into the details in the next days.

I assume that this worked when `RcParams` was just inheriting from dict (not tested).

Anyway, in the long term, we should aim at just inheriting from `MutableMapping`.

**Comment by anntzer** [comment]
> Could we just implement __len__ as a quicker fix? I am curious of this ever worked....

I think it's a bug in CPython: https://bugs.python.org/issue35063

**Comment by timhoffm** [comment]
Not having fully read the bug, but implementing `__len__` would work around this, right?

**Comment by anntzer** [comment]
Yes.
The bug is that RcParams() should have raised an exception given than `__len__` is a not-overridden abstractmethod.

**Comment by timhoffm** [comment]
Closing as the current problem is fixed by #12630.

Still, I would like to get rid of the dict-inheritance in the the long run (#12577).
